Assets/Scripts/MovePlayer.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class MovePlayer : MonoBehaviour
+{
+    #region SerializeFields
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a slow movement (through sand, etc.).")]
+    private float slowMoveTime = .75f;
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a single movement.")]
+    private float moveTime = .5f;
+    [SerializeField]
+    [Tooltip("Threshold to stop movement coroutine and snap to position.")]
+    private float movementMargin = .05f;
+    [SerializeField]
+    [Tooltip("Point from which to raycast to check collision before moving.")]
+    private Transform collisionRaycastPoint;
+    [SerializeField]
+    [Tooltip("Layer(s) the player collides with when moving.")]
+    private LayerMask collisionLayerMask;
+    [SerializeField]
+    [Tooltip("Layer that slows the player down when moving.")]
+    private LayerMask slowTerrainLayerMask;
+    [SerializeField]
+    [Tooltip("Tag of terrain that slows the player down when moving.")]
+    private string slowTerrainTag;
+    #endregion
+
+    /// <summary>
+    /// Defines directions in an enum for legible array reference
+    /// </summary>
+    private enum directions { Down, Up, Left, Right};
+
+    /// <summary>
+    /// The animator attached to the player
+    /// </summary>
+    private Animator playerAnimator;
+    /// <summary>
+    /// Is the player allowed to move? Will be false if paused, in a cutscene, etc.
+    /// </summary>
+    private bool canMove = true;
+    /// <summary>
+    /// Has the player received input to move?
+    /// </summary>
+    private bool shouldMove;
+    /// <summary>
+    /// Is the player in slow terrain?
+    /// </summary>
+    private bool slowMove;
+    /// <summary>
+    /// Is the movement coroutine currently running?
+    /// </summary>
+    private bool moving;
+    /// <summary>
+    /// Which direction is the player trying to move?
+    /// </summary>
+    private directions currentDirection;
+    /// <summary>
+    /// An array for quick reference of directions, in conjunction with the enum
+    /// This prevents the need to write four separate coroutines
+    /// </summary>
+    private Vector3[] movementVectors;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        playerAnimator = GetComponent<Animator>();
+        movementVectors = new Vector3[4] { Vector2.down, Vector2.up, Vector2.left, Vector2.right };
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        GetInput();
+    }
+
+    private void FixedUpdate()
+    {
+        if (canMove && shouldMove && !moving)
+            StartCoroutine(MovementRoutine());
+    }
+
+    private void OnTriggerEnter2D(Collider2D collision)
+    {
+        if (collision.gameObject.tag == slowTerrainTag)
```

Comment:
That ended up being the solution I reached for getting the slow terrain to work. The nature of 2D colliders caused a lot of problems for me in that, and if there's a better way to do it, I'd be glad to change it.

Reply:
To clarify, I couldn't find any way to compare the layer of the collided object outside of a bitwise operation I found in a Unity forum post, which I didn't want to use because I didn't know exactly what it does. The layer is for the raycast, the tag is for the trigger enter. Again, if we can find a better solution that remains functional, I'll gladly drop the tag.